pkg/destroy: data/aws: delete instance profiles even if they are detached

eparis opened the pull request:
Today we find instance profiles that we need to delete through either the
associated role or the associated instance. We find those objects via aws
tags. If those objects have been deleted we are unable to find the instance
profiles because those are not tagged.

Since we now embed both cluster id inside the name this PR adds the ability to find those object and to delete them even if the roles and instances have been deleted.

eparis commented:
@wking get ready to throw up when you see what I did to your lovely destroy code   :)

wking commented:
With cluster IDs in the profile names, new clusters won't be bothered by any previous instance profiles leaked by buggy reapers (the #1174 issue).  So I'm in favor of adding the cluster ID here (and maybe removing the cluster name if we are concerned about name-length limits?), but I don't think we need to change the destroy code.

staebler commented on pkg/destroy/aws/aws.go:
```diff
@@ -184,6 +184,51 @@ func (o *ClusterUninstaller) Run() error {
+// deleted artifacts behind our back.
+func (o *ClusterUninstaller) deleteInstanceProfileHack(awsSession *session.Session) error {
+	o.Logger.Debug(fmt.Sprintf("%v", o))
+	clusterID := ""
```
We use the term `clusterName` instead of `clusterID` for this elsewhere in the installer code.

staebler commented on pkg/destroy/aws/aws.go:
```diff
@@ -184,6 +184,51 @@ func (o *ClusterUninstaller) Run() error {
+	if openshiftClusterID == "" {
+		return fmt.Errorf("unable to determine openshiftClusterID in deleteInstanceProfileHack()")
+	}
+	masterARN := fmt.Sprintf("arn:aws:iam::123456789012:instance-profile/%s-master-profile-%s", clusterID, openshiftClusterID)
```
My preference would be to have a `deleteIAMInstanceProfileByName` function that is passed the name of the profile rather than making up an ARN with a fake account ID that the caller is trusting won't actually be used.

eparis commented:
updated to address review comments.

I've also changed my mind, I don't think this is a 'hack'. It is cleaning up what we created. We know we created it, we should clean it up.

wking commented:
> I've also changed my mind, I don't think this is a 'hack'. It is cleaning up what we created. We know we created it, we should clean it up.

I think once another reaper leans in and deletes our instances and IAM roles, it's assuming responsibility for reaping instance profiles too.  I don't think we need to support step 3 in:

1. User neglects to call `destroy cluster`
2. Reaper deletes most resources.
3. User reassembles `metadata.json` and calls `destroy cluster`.

Are users who pass step 1 likely to take step 3 when creating a new cluster with the same name no longer has resource conflicts (because of your appended cluster ID)?

eparis commented:
I'm not sure why #3 is 'User reassembles`.  All the user needs to do is call destroy_cluster in #3 for this PR to be helpful.

While I agree that if something else starts to clean up a cluster it probably should clean up the entire cluster, I say that if the installer creates something the installer should be able to destroy it. A customer could (less likely admittedly) call `destroy cluster` and have an AWS API malfunction at just the wrong time leaving them in the same state. Running `destroy cluster` a second time can, and I say should, continue to finish the cleanup.

wking commented:
> A customer could (less likely admittedly) call destroy cluster and have an AWS API malfunction at just the wrong time leaving them in the same state.

We currently order removal so roles and instances are not removed before the instance profile.  So "network hiccup" is not sufficient, you'd need "AWS API returns no associated instance-profiles when that instance profile actually did exist" to get into trouble here.

eparis commented:
I just need to remove the role, instance profile removal will fail. remove the instance, hickup, stuck. No?

wking commented:
> I just need to remove the role, instance profile removal will fail. remove the instance, hickup, stuck. No?

We only attempt to [remove roles][1] after successfully [removing associated instance profiles][2].  [Same for instances][3].  So how do you get the role and instance removed first, except via a buggy external reaper?

[1]: https://github.com/openshift/installer/blob/8dd7e3f3757968c32641b8923461a470c1355ccf/pkg/destroy/aws/aws.go#L1146
[2]: https://github.com/openshift/installer/blob/8dd7e3f3757968c32641b8923461a470c1355ccf/pkg/destroy/aws/aws.go#L1117
[3]: https://github.com/openshift/installer/blob/8dd7e3f3757968c32641b8923461a470c1355ccf/pkg/destroy/aws/aws.go#L523

eparis commented:
/retest

eparis commented:
/retest

eparis commented:
/retest

eparis commented:
/rest

staebler commented on pkg/destroy/aws.go:
```diff
@@ -14,9 +14,10 @@ func NewAWS(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (Destroy
+		Filters:   filters,
+		Region:    metadata.ClusterPlatformMetadata.AWS.Region,
+		Logger:    logger,
+		ClusterID: metadata.ClusterID,
```
This should us `metadata.InfraID` now.

abhinavdahiya commented:
ping @wking can you take another look at the PR?

wking commented on pkg/destroy/aws.go:
```diff
@@ -14,9 +14,10 @@ func NewAWS(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (Destroy
+		Filters:   filters,
+		Region:    metadata.ClusterPlatformMetadata.AWS.Region,
+		Logger:    logger,
+		ClusterID: metadata.InfraID,
```
This will cause a trivial conflict with #1365, which is a higher-priority bugfix.

wking commented on pkg/destroy/aws/aws.go:
```diff
@@ -1132,12 +1163,25 @@ func deleteIAM(session *session.Session, arn arn.ARN, logger logrus.FieldLogger)
+		}
+		return err
+	}
+	logger.WithField("name", name).Info("Deleted")
```
nit: I'd rather have the caller set this context, like we do [here][1], etc.  The caller can chose which of the information available to it should be logged.  Functions can add additional logging context as new information become available, but they shouldn't write their arguments directly into the logging context.

[1]: https://github.com/openshift/installer/blob/5e34eedac7037178c9d01a0a16cde46d361fc873/pkg/destroy/aws/aws.go#L466

wking commented:
I left two minor nits inline.  Otherwise, this looks fine to me, and I'm ok with it landing (with or without the nits getting addressed), although I still don't feel like we need it ;).

abhinavdahiya commented:
> I left two minor nits inline. Otherwise, this looks fine to me, and I'm ok with it landing (with or without the nits getting addressed), although I still don't feel like we need it ;).

the nits don't look like blockers.
and i think the destroyer should try it **best** to cleanup the resources from AWS for the cluster, so I think this is useful change. 

/lgtm

PS: It's sad that the instance profiles cannot be tagged :(  .
